R/stackingutilities.R

```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+# tidying patterns that are applied to grids in Displayr
+# which mean that text labels may not ititially match. That is
+# labels on the code frames of grid questions commonly have
+# aditional common prefix and suffix text removed.
+prepareTextVariableLabelsForStackingWithGrids <- function(text, categorical.question) {
+    categorical.codeframe.labels <- names(attr(categorical.question, "codeframe"))
+    categorical.secondary.codeframe.labels <- names(attr(categorical.question, "secondarycodeframe"))
+    text.labels <- colnames(text)
+
+    if (!any(text.labels %in% categorical.codeframe.labels) && !any(text.labels %in% categorical.secondary.codeframe.labels)) {
+        # No matches for the text labels anywhere
+        # Remove common prefixes
+        text.labels <- flipFormat::TidyLabels(text.labels)
+    }
+
+    if (!any(text.labels %in% categorical.codeframe.labels) && !any(text.labels %in% categorical.secondary.codeframe.labels)) {
+        # Still No matches for the text labels anywhere
+        # Remove common suffixes
+        text.labels <- removeCommonSuffixText(text.labels)
+    }
+
+    if (!any(text.labels %in% categorical.codeframe.labels) && !any(text.labels %in% categorical.secondary.codeframe.labels)) {
+        stop(paste0("Unable to match the labels from the Text variables to the labels of ", 
+                    attr(categorical.question, "question"), 
+                    ". Please modify the labels so that the text variables may be matched."))
+    }
+
+    # Update labels and return
+    attr(text, "codeframe") <- fakeCodeFrame(text.labels)
+    colnames(text) <- text.labels
+    text
+}
+
+# Generates a vector which mimics the "codeframe"
+# attribute for a Displayr variable/variable set.
+# This is an integer vector named according to
+# 'labels'.
+fakeCodeFrame <- function(labels) {
+    fake.codes <- 1:length(labels)
+    names(fake.codes) <- labels
+    fake.codes
+}
+
+# Function to identify and remove the common text
+# from ends of the supplied labels. This is useful
+# when trying to match labels to labels from
+# variable sets in Displayr where redundant text
+# is often removed from labels.
+#' @importFrom stringi stri_reverse
+removeCommonSuffixText <- function(labels) {
+    reversed.labels <- stri_reverse(labels)
+    common <- longestCommonPrefix(reversed.labels)
+    pattern <- paste0("^", common)
+    reversed.labels <- gsub(pattern, "", reversed.labels)
+    labels <- stri_reverse(reversed.labels)
+}
+
+# From https://stackoverflow.com/questions/28273716/r-implementation-for-finding-the-longest-common-starting-substrings-in-a-set-of
+longestCommonPrefix <- function(x) {
+    # sort the vector
+    x <- sort(x)
+    # split the first and last element by character
+    d_x <- strsplit(x[c(1,length(x))],"")
+    # search for the first not common element and so, get the last matching one
+    der_com <- match(FALSE, do.call("==", d_x)) - 1
```

Comment:
This will throw an uninformative warning when used. Suggest changing this function.
```
longestCommonPrefix(c("foobarbaz", "foobarwaz", "foobazzzzz"))
```

Reply:
```suggestion
longestCommonPrefix <- function(x) {
    # sort the vector
    x <- sort(x)
    # split the first and last element by character
    chars <- strsplit(x[c(1, length(x))], "")
    # Find shortest string length
    n.shortest <- length(chars[[1L]])
    # search for the first not common element and so, get the last matching one
    if (length(chars[[2L]]) != n.shortest)
        chars[[2L]] <- chars[[2L]][1:n.shortest]
    common.prefix.length <- which.max(!chars[[1]] == chars[[2]]) - 1L
    if (common.prefix.length == 0L)
        return(character(0L))
    substr(x[1], 1L, common.prefix.length)
}
```